Add functions to set in game Cursor Color for client

Proxy-99 opened the pull request:
add new client functions
```
setCursorColor(0,255,30,255)
setCursorAlpha(100)
r,g,b,a = getCursorColor()
print(r)
print(g)
print(b)
print(a)
```
part of #3555

Proxy-99 committed:
cursor color function

Proxy-99 committed:
fix wrong swap

Proxy-99 committed:
fixes swap parameters

Proxy-99 committed:
fixes , add getCursorColor

Proxy-99 committed:
on reconnect reset values

tederis commented on Client/core/CGUI.cpp:
```diff
@@ -431,7 +425,12 @@ void CLocalGUI::SetMainMenuVisible(bool bVisible)
-            pGUI->SetCursorAlpha(pGUI->GetCurrentServerCursorAlpha());
+        {
+            float r,g,b,a;
+            pGUI->GetCursorColor(r,g,b,a);
```
Please, add spaces between arguments.

tederis commented on Client/gui/CGUI_Impl.cpp:
```diff
@@ -502,9 +502,29 @@ void CGUI_Impl::SetCursorAlpha(float fAlpha, bool bOnlyCurrentServer)
 
+void CGUI_Impl::SetCursorColor(float r, float g, float b, float alpha)
+{
+    CEGUI::MouseCursor::getSingleton().setColor(r,g,b,alpha);
```
Please, add spaces between arguments.

tederis commented on Client/gui/CGUI_Impl.h:
```diff
@@ -145,6 +145,8 @@ class CGUI_Impl : public CGUI, public CGUITabList
     void        SetCurrentServerCursorAlpha(float fAlpha);
     float       GetCurrentServerCursorAlpha();
+    void        SetCursorColor(float r, float g, float b, float alpha);
+    void        GetCursorColor(float &r,float &g,float &b,float &alpha);
```
Please, add spaces between arguments.

tederis commented on Client/gui/CGUI_Impl.h:
```diff
@@ -311,7 +313,10 @@ class CGUI_Impl : public CGUI, public CGUITabList
+    float                 m_fCurrentServerCursorAlpha =1.0f;
+    float                 m_fCurrentServerCursorRed= 1.0f;
+    float                 m_fCurrentServerCursorGreen=1.0f;
+    float                 m_fCurrentServerCursorBlue= 1.0f;
```
Invalid formatting. Please, add spaces in front of the 'equals' sign.

tederis commented on Client/mods/deathmatch/logic/luadefs/CLuaClientDefs.cpp:
```diff
@@ -80,3 +82,18 @@ bool CLuaClientDefs::IsCapsLockEnabled()
+{
+    float r, g, b, alpha;
+    CGUI* pGUI = g_pCore->GetGUI();
+    pGUI->GetCursorColor(r,g,b,alpha);
```
Please, add spaces between arguments.

tederis commented:
Will the colors be reset after a disconnect?

Proxy-99 commented:
> Will the colors be reset after a disconnect?

I will try to put the function within onClientPlayerQuit event

Proxy-99 committed:
Merge branch 'master' into cursor

tederis commented:
> > Will the colors be reset after a disconnect?
> 
> I will try to put the function within onClientPlayerQuit event

There is the special place for it: `CClientGame::Event_OnIngame`

Proxy-99 committed:
review fixes

Proxy-99 committed:
bug fix when using setCursorAlpha and start resource using F8 won't refresh i

setCursorAlpha(150) when you open up the F8 and start the resource it won't refresh the image due to a check if F8 console is open it or not

Proxy-99 committed:
refactor setCursorAlpha

Proxy-99 commented:
@tederis 
I have refactor part of setCursorAlpha() function to set the color correctly
 
here is my test 
[https://youtu.be/svReSqdAcdg?si=ElnAsBf23XYHhrue](https://youtu.be/svReSqdAcdg?si=ElnAsBf23XYHhrue)
seems all good

lopezloo commented:
I don't think it should be possible to set this for pause menu since it could be abused to hide cursor.

Proxy-99 committed:
menu cursor color fix

Proxy-99 committed:
reset color correctly

Proxy-99 committed:
clean up

Proxy-99 commented:
> I don't think it should be possible to set this for pause menu since it could be abused to hide cursor.

now fixed
when you open up the main menu it will reset the color to the deafult one
when you close it it will set the server color back again
when you disconnect it will reset the values and color to deafult

Proxy-99 committed:
bug fix of values

Proxy-99 committed:
Merge branch 'master' into cursor

Proxy-99 committed:
float fixes

tederis commented on Client/gui/CGUI_Impl.h:
```diff
@@ -145,6 +145,10 @@ class CGUI_Impl : public CGUI, public CGUITabList
+    void        SetCursorColor(float r, float g, float b, float alpha);
+    void        GetCursorColor(float &r, float &g, float &b, float &alpha);
+    void        ResetMenuCursorColor();
+    void        RestCursorColor();
```
Did you mean `ResetCursorColor`?

tederis commented on Client/mods/deathmatch/logic/CClientGame.cpp:
```diff
@@ -3424,6 +3424,9 @@ void CClientGame::Event_OnIngame()
     g_pGame->ResetAlphaTransparencies();
     g_pGame->ResetModelTimes();
 
+    //reset cursor color
```
Just to be consistent with the other code. Please, add a space after two slashes and start a comment with first letter in upper case. Here and in other places as well.

tederis commented on Client/mods/deathmatch/logic/luadefs/CLuaClientDefs.cpp:
```diff
@@ -80,3 +82,20 @@ bool CLuaClientDefs::IsCapsLockEnabled()
+
+bool CLuaClientDefs::SetCursorColor(float r, float g, float b, float alpha)
+{
+    if (!g_pCore->IsMenuVisible())
```
This function must not reject the color. See how this is implemented for `setCursorAlpha`:
```
if (!m_pCore->IsMenuVisible() && !m_pCore->GetConsole()->IsVisible())
    m_pGUI->SetCursorAlpha(fAlpha, true);
else
    m_pGUI->SetCurrentServerCursorAlpha(fAlpha);
```

TracerDS commented on Client/mods/deathmatch/logic/luadefs/CLuaClientDefs.cpp:
```diff
@@ -80,3 +82,20 @@ bool CLuaClientDefs::IsCapsLockEnabled()
+    return false;   
+}
+
+CLuaMultiReturn<float, float, float, float> CLuaClientDefs::GetCursorColor()
```
`noexcept`?

TracerDS commented on Client/gui/CGUI_Impl.cpp:
```diff
@@ -496,12 +496,42 @@ bool CGUI_Impl::IsCursorEnabled()
+    m_fCurrentServerCursorGreen = 255.0f;
+    m_fCurrentServerCursorBlue = 255.0f;
+    m_fCurrentServerCursorAlpha = 1.0f;
+}
```
`noexcept`?

TracerDS commented on Client/core/CGUI.cpp:
```diff
@@ -431,7 +425,12 @@ void CLocalGUI::SetMainMenuVisible(bool bVisible)
+            pGUI->GetCursorColor(r, g, b, a); //restore the cursor color when we close the main menu
+            pGUI->SetCursorColor(r, g, b, a);
+        }
+           
```
whats the point of this code?

Proxy-99 replied:
when we close the main menu it will retrieve the cursor color and set it again

TracerDS replied:
Why do we set it again? Wouldnt that change nothing?

Proxy-99 replied:
if you open the main menu the cursor color will be white 
if we close the menu again it will return the color again

it just part of a some code in Client\core\CMessageLoopHook.cpp
I think some sort of keybind handler when we press escape 
so if we press escape to close the menu it will call 
`void CLocalGUI::SetMainMenuVisible(bool bVisible)`
then we get to the setCursorColor

```
        // Pass escape keyup to onClientKey
        if (uMsg == WM_KEYUP && wParam == VK_ESCAPE)
        {
            g_pCore->GetKeyBinds()->TriggerKeyStrokeHandler("escape", uMsg == WM_KEYDOWN, true);
            return true;
        }

        // Suppress auto repeat of escape and console toggle keys
        if ((uMsg == WM_KEYDOWN || uMsg == WM_CHAR) && (wParam == VK_ESCAPE || wParam == VK_F8 || wParam == '`'))
        {
            bool bFirstHit = (lParam & 0x40000000) ? false : true;
            if (!bFirstHit)
                return true;
        }

        // Slightly hacky way of suppressing escape character when console is closed with escape key
        if (uMsg == WM_CHAR && wParam == VK_ESCAPE)
        {
            bool bTemp = ms_bIgnoreNextEscapeCharacter;
            ms_bIgnoreNextEscapeCharacter = false;
            if (bTemp)
                return true;
        }

        if (CKeyBinds::IsFakeCtrl_L(uMsg, wParam, lParam))
            return true;

        bool       bWasCaptureKey = false;
        CMainMenu* pMainMenu = g_pCore->GetLocalGUI()->GetMainMenu();
        if (pMainMenu)
        {
            CSettings* pSettings = pMainMenu->GetSettingsWindow();
            if (pSettings)
            {
                if (uMsg == WM_KEYDOWN && wParam == VK_ESCAPE && GetJoystickManager()->IsCapturingAxis())
                {
                    GetJoystickManager()->CancelCaptureAxis(true);
                    return true;
                }
                bWasCaptureKey = (pSettings->IsCapturingKey() && pSettings->ProcessMessage(uMsg, wParam, lParam));
                if (!bWasCaptureKey)
                {
                    // If Escape is pressed and we're playing ingame, we show/hide the menu
                    if (uMsg == WM_KEYDOWN && wParam == VK_ESCAPE && g_pCore->IsConnected())
                    {
                        // Hide the console
                        CConsoleInterface* pConsole = g_pCore->GetConsole();
                        if (pConsole->IsVisible())
                        {
                            ms_bIgnoreNextEscapeCharacter = true;
                            pConsole->SetVisible(false);

                            CGUI* pGUI = g_pCore->GetGUI();
                            float r, g, b, alpha;

                            if (!g_pCore->IsMenuVisible())
                            {
                                pGUI->GetCursorColor(r, g, b, alpha);
                                pGUI->SetCursorColor(r, g, b, alpha);
                            }
                      
                            return true;
                        }

                        // The mainmenu makes sure it isn't hidden if UseIngameButtons == false
                        if (!CCore::GetSingleton().IsOfflineMod())
                        {
                            if (g_pCore->GetKeyBinds()->TriggerKeyStrokeHandler("escape", uMsg == WM_KEYDOWN, true))
                            {
                                // Stop chat input
                                if (CLocalGUI::GetSingleton().IsChatBoxInputEnabled())
                                {
                                    CLocalGUI::GetSingleton().SetChatBoxInputEnabled(false);
                                    return true;
                                }

                                CLocalGUI::GetSingleton().SetMainMenuVisible(!CLocalGUI::GetSingleton().IsMainMenuVisible());
                            }
                            return true;
                        }
                    }
                    else if (uMsg == WM_KEYDOWN && wParam == VK_ESCAPE && !g_pCore->IsConnected())
                    {
                        // If Escape is pressed and we're not playing ingame, hide certain windows
                        CLocalGUI::GetSingleton().GetMainMenu()->OnEscapePressedOffLine();
                    }
```

TracerDS replied:
> if you open the main menu the cursor color will be white if we close the menu again it will return the color again
> 
> it just part of a some code in Client\core\CMessageLoopHook.cpp I think some sort of keybind handler when we press escape so if we press escape to close the menu it will call `void CLocalGUI::SetMainMenuVisible(bool bVisible)` then we get to the setCursorColor

I mean, `GetCursorColor` will return current color. Doing `SetCursorColor` with the same values doesnt make any sense.
If without that the code works correctly then something is horribly wrong behind the scene

Proxy-99 replied:
fixed in
(https://github.com/multitheftauto/mtasa-blue/pull/3558/commits/127efaa1b0192a7c3f14754b55d4789df742c4f1)

Proxy-99 committed:
tederis review fixes

Fernando-A-Rocha commented:
What will happen if the player has the main menu open and the client script calls `setCursorColor`? Does it change the color while the menu is indeed open (it shouldn't imo)?

Proxy-99 committed:
review 2 tracer

Proxy-99 commented:
> What will happen if the player has the main menu open and the client script calls `setCursorColor`? Does it change the color while the menu is indeed open (it shouldn't imo)?

it does not change the color but it change the members variables values so when you close the menu the color will be set and up to date

Proxy-99 committed:
small fix for useless if condition

Proxy-99 commented:
for testing result see this @TracerDS @Fernando-A-Rocha 
[https://youtu.be/5mGd4oKjTGM](https://youtu.be/5mGd4oKjTGM)
you can test with this code

```lua
setCursorColor(0,255,30,1)
setCursorAlpha(100)
r,g,b,a = getCursorColor()
print(r)
print(g)
print(b)
print(a)

```

Proxy-99 committed:
simplify

Proxy-99 committed:
typos

FileEX commented on Client/core/CMessageLoopHook.cpp:
```diff
@@ -295,9 +295,11 @@ LRESULT CALLBACK CMessageLoopHook::ProcessMessage(HWND hwnd, UINT uMsg, WPARAM w
+                            if (!g_pCore->IsMenuVisible())
+                            {
+                                pGUI->RestoreCurrentServerCursorColor();
+                            }
```
You can remove the braces

FileEX commented on Client/gui/CGUI_Impl.cpp:
```diff
@@ -496,12 +496,47 @@ bool CGUI_Impl::IsCursorEnabled()
+    m_fCurrentServerCursorAlpha = alpha;
+}
+
+void CGUI_Impl::RestoreCurrentServerCursorColor()
```
noexcept

FileEX commented on Client/gui/CGUI_Impl.h:
```diff
@@ -145,6 +145,11 @@ class CGUI_Impl : public CGUI, public CGUITabList
+    void        SetCursorColor(float r, float g, float b, float alpha) noexcept;
+    void        GetCursorColor(float& r, float& g, float& b, float& alpha) noexcept;
+    void        ResetMenuCursorColor() noexcept;
+    void        RestoreCurrentServerCursorColor();
```
noexcept

FileEX commented on Client/mods/deathmatch/logic/luadefs/CLuaClientDefs.cpp:
```diff
@@ -80,3 +82,23 @@ bool CLuaClientDefs::IsCapsLockEnabled()
     return ((::GetKeyState(VK_CAPITAL) & 0x0001) != 0);
 }
+
+bool CLuaClientDefs::SetCursorColor(float r, float g, float b, float alpha)
```
noexcept

FileEX commented on Client/mods/deathmatch/logic/luadefs/CLuaClientDefs.h:
```diff
@@ -17,6 +17,8 @@ class CLuaClientDefs : public CLuaDefs
 {
 public:
     static void LoadFunctions();
+    static bool SetCursorColor(float r, float g, float b, float alpha);
```
noexcept

FileEX commented on Client/sdk/gui/CGUI.h:
```diff
@@ -126,6 +126,11 @@ class CGUI
+    virtual void        GetCursorColor(float &r, float &g, float &b, float &alpha) = 0;
+    virtual void        ResetMenuCursorColor() = 0;
+    virtual void        ResetCursorColor(float r, float g, float b, float alpha) = 0;
+    virtual void        RestoreCurrentServerCursorColor() = 0;
```
noexcept also

FileEX commented on vendor/cegui-0.4.0-custom/src/CEGUIMouseCursor.cpp:
```diff
@@ -149,12 +149,18 @@ void MouseCursor::offsetPosition(const Point& offset)
+/*************************************************************************
+    Set the mouse cursor's color
+*************************************************************************/
+void MouseCursor::setColor(float r, float g, float b, float alpha)
```
noexcept?

FileEX commented on vendor/cegui-0.4.0-custom/src/CEGUIMouseCursor.cpp:
```diff
@@ -149,12 +149,18 @@ void MouseCursor::offsetPosition(const Point& offset)
     Set the mouse cursor's alpha
 *************************************************************************/
-void MouseCursor::setAlpha(float alpha)
+void MouseCursor::setAlpha(float r, float g, float b, float alpha)
```
noexcept?

FileEX commented on vendor/cegui-0.4.0-custom/include/CEGUIMouseCursor.h:
```diff
@@ -187,7 +187,8 @@ class CEGUIEXPORT MouseCursor : public EventSet, public Singleton<MouseCursor>
     */
-    void    setAlpha(float alpha);
+    void    setAlpha(float r, float g, float b, float alpha);
+    void    setColor(float r, float g, float b, float alpha);
```
noexcept?

FileEX commented on Client/mods/deathmatch/logic/luadefs/CLuaClientDefs.cpp:
```diff
@@ -80,3 +82,23 @@ bool CLuaClientDefs::IsCapsLockEnabled()
+    else
+    {
+      g_pCore->GetGUI()->ResetCursorColor(r, g, b, alpha); // Force values to be updated
+    }
```
You can remove the braces

Proxy-99 committed:
Merge branch 'master' into cursor

Proxy-99 committed:
review fixes

Proxy-99 committed:
noexcept

Proxy-99 committed:
Merge branch 'master' into cursor

Proxy-99 committed:
Merge branch 'master' into cursor

Fernando-A-Rocha reviewed (approved):
Good. I tested: it works fine.

![image](https://github.com/user-attachments/assets/c9b7fabe-2a17-48e2-a107-71caaaa8644a)

```lua
addCommandHandler("cursorcolor", function(cmd, arg)
    if arg == "green" then
        setCursorColor(0,255,30,1)
    else
        setCursorColor(255,255,255,1)
    end
    iprint(getCursorColor())
end)
```

Proxy-99 committed:
Merge branch 'master' into cursor

Fernando-A-Rocha commented on Client/mods/deathmatch/logic/luadefs/CLuaClientDefs.cpp:
```diff
@@ -23,7 +23,9 @@ void CLuaClientDefs::LoadFunctions()
                                                                              {"isMTAWindowFocused", ArgumentParser<IsMTAWindowFocused>},
-                                                                             {"isCapsLockEnabled", ArgumentParser<IsCapsLockEnabled>}};
+                                                                             {"isCapsLockEnabled", ArgumentParser<IsCapsLockEnabled>},
+                                                                             {"setCursorColor", ArgumentParser<SetCursorColor>},
+                                                                             {"getCursorColor", ArgumentParser<GetCursorColor>}};
```
Can you also add `resetCursorColor` to make it normal (white) again?

FileEX replied:
It doesn't make sense. Just like there is no ``resetCursorAlpha`` function, there shouldn't be a ``resetCursorColor`` because it would essentially be just a call to ``setCursorColor(255, 255, 255)``

Fernando-A-Rocha replied:
Fair enough. But the wiki should properly document the default color
And perhaps if no arguments are passed to the set function it resets to default?

FileEX replied:
Arguments should be optional
```lua
setCursorColor( [ int r = 255, int g = 255, int b = 255 ] )
```
so, yes

```lua
setCursorColor()
```
should set default white color

Proxy-99 replied:
I have added the setCursorColor() that was suggested

Fernando-A-Rocha commented on Client/mods/deathmatch/logic/luadefs/CLuaClientDefs.cpp:
```diff
@@ -80,3 +82,20 @@ bool CLuaClientDefs::IsCapsLockEnabled()
+bool CLuaClientDefs::SetCursorColor(float r, float g, float b, float alpha) noexcept
+{
+    if (!g_pCore->IsMenuVisible())
+        g_pCore->GetGUI()->SetCursorColor(r, g, b, alpha);
```
There should be no alpha argument in setCursorColor. [setCursorAlpha](https://wiki.multitheftauto.com/wiki/SetCursorAlpha) already exists, and it is range 0-255, not 0-1. If you really believe it should exist in setCursorColor, then fix the value range to 0-255.

Proxy-99 replied:
having alpha makes more sense to set color and alpha at the same time while using setCursorAlpha is up to u
I will make it optional argument

Fernando-A-Rocha replied:
Sure make optional arg

Fernando-A-Rocha replied:
But for the developer, alpha should be 0-255

Proxy-99 replied:
I think having range between 0 and 1 is way readable and predictable than the format with 255 but maybe will affect some codes that rely on 255 format when getting the alpha they should consider handle it correctly

Fernando-A-Rocha replied:
> I think having range between 0 and 1 is way readable and predictable than the format with 255 but maybe will affect some codes that rely on 255 format when getting the alpha they should consider handle it correctly

I disagree, all other transparency related code in MTA is with 0-255 range. For consistency we shouldn't change that

Fernando-A-Rocha commented on Client/mods/deathmatch/logic/luadefs/CLuaClientDefs.cpp:
```diff
@@ -80,3 +82,20 @@ bool CLuaClientDefs::IsCapsLockEnabled()
+{
+    float r, g, b, alpha;
+    g_pCore->GetGUI()->GetCursorColor(r, g, b, alpha);
+    return {r, g, b, alpha};
```
See the comment above. We already have [getCursorAlpha](https://wiki.multitheftauto.com/wiki/GetCursorAlpha) to return the alpha value...

Proxy-99 replied:
no need to call two functions getting alpha through one function a way easier

Proxy-99 committed:
Merge branch 'master' into cursor

Proxy-99 committed:
Merge branch 'master' into cursor

Proxy-99 committed:
make arguments optional

reset cursor via setCursorColor()

Proxy-99 committed:
Merge branch 'master' into cursor

znvjder reviewed (commented):
Can you make please range (0-255) also for alpha color?

Proxy-99 commented:
> Can you make please range (0-255) also for alpha color?

I will rewrite it soon

Fernando-A-Rocha commented:
Ok, good :D

Proxy-99 committed:
alpha to work with range 0 to 255

Proxy-99 commented:
Since I think it looks all good now and reviews by many can you merge this @botder @tederis

Fernando-A-Rocha reviewed (approved):
👏

FileEX commented on vendor/cegui-0.4.0-custom/src/CEGUIMouseCursor.cpp:
```diff
@@ -149,12 +149,18 @@ void MouseCursor::offsetPosition(const Point& offset)
 {
-    d_colourRect.setColours(colour(1.0f, 1.0f, 1.0f, alpha));
+    d_colourRect.setColours(colour(r/255.f, g/255.f, b/255.f, alpha));
 }
```
This function should be removed because it no longer only changes the alpha and does exactly the same thing as setColor. Additionally, if the setColours function is not noexcept, then setColour should also not be noexcept

FileEX commented on vendor/cegui-0.4.0-custom/src/CEGUIMouseCursor.cpp:
```diff
@@ -149,12 +149,18 @@ void MouseCursor::offsetPosition(const Point& offset)
+/*************************************************************************
+    Set the mouse cursor's color
+*************************************************************************/
+void MouseCursor::setColor(float r, float g, float b, float alpha) noexcept
+{
+    d_colourRect.setColours(colour(r/255.f, g/255.f, b/255.f, alpha/255.f));
+}
```
If the setColours function is not noexcept, then setColour should also not be noexcept

FileEX commented on Client/gui/CGUI_Impl.h:
```diff
@@ -311,7 +316,10 @@ class CGUI_Impl : public CGUI, public CGUITabList
+    float                 m_fCurrentServerCursorAlpha = 255.0f;
+    float                 m_fCurrentServerCursorRed = 255.0f;
+    float                 m_fCurrentServerCursorGreen = 255.0f;
+    float                 m_fCurrentServerCursorBlue = 255.0f;
```
Hungarian notation

FileEX commented on Client/gui/CGUI_Impl.h:
```diff
@@ -145,6 +145,11 @@ class CGUI_Impl : public CGUI, public CGUITabList
+    void        GetCursorColor(float& r, float& g, float& b, float& alpha) noexcept;
+    void        ResetMenuCursorColor() noexcept;
+    void        RestoreCurrentServerCursorColor() noexcept;
+    void        ResetCursorColor(float r, float g, float b, float alpha) noexcept;
```
Please use ``override``

FileEX commented on Client/gui/CGUI_Impl.cpp:
```diff
@@ -496,12 +496,47 @@ bool CGUI_Impl::IsCursorEnabled()
+    alpha = m_fCurrentServerCursorAlpha;
+}
+
+void CGUI_Impl::ResetMenuCursorColor() noexcept
+{
+    CEGUI::MouseCursor::getSingleton().setColor(255.0f, 255.0f, 255.0f, 255.0f);
+}
```
If a function calls another function that is not noexcept, then it should also not be noexcept, because this contradicts the assumptions of that keyword

FileEX commented on Client/gui/CGUI_Impl.cpp:
```diff
@@ -496,12 +496,47 @@ bool CGUI_Impl::IsCursorEnabled()
+    m_fCurrentServerCursorAlpha = alpha;
+}
+
+void CGUI_Impl::RestoreCurrentServerCursorColor() noexcept
+{
+    CEGUI::MouseCursor::getSingleton().setColor(m_fCurrentServerCursorRed, m_fCurrentServerCursorGreen, m_fCurrentServerCursorBlue, m_fCurrentServerCursorAlpha);
+}
```
If a function calls another function that is not noexcept, then it should also not be noexcept, because this contradicts the assumptions of that keyword

FileEX commented on Client/gui/CGUI_Impl.cpp:
```diff
@@ -496,12 +496,47 @@ bool CGUI_Impl::IsCursorEnabled()
 
+void CGUI_Impl::SetCursorColor(float r, float g, float b, float alpha) noexcept
+{
+    CEGUI::MouseCursor::getSingleton().setColor(r, g, b, alpha);
```
If a function calls another function that is not noexcept, then it should also not be noexcept, because this contradicts the assumptions of that keyword

FileEX commented on Client/gui/CGUI_Impl.cpp:
```diff
@@ -496,12 +496,47 @@ bool CGUI_Impl::IsCursorEnabled()
 void CGUI_Impl::SetCursorAlpha(float fAlpha, bool bOnlyCurrentServer)
 {
-    CEGUI::MouseCursor::getSingleton().setAlpha(fAlpha);
+    CEGUI::MouseCursor::getSingleton().setAlpha(m_fCurrentServerCursorRed, m_fCurrentServerCursorGreen, m_fCurrentServerCursorBlue, fAlpha);
```
setColour

FileEX commented on Client/mods/deathmatch/logic/CClientGame.cpp:
```diff
@@ -3443,6 +3443,9 @@ void CClientGame::Event_OnIngame()
     g_pGame->SetWeaponRenderEnabled(true);
+  
+    // Reset cursor color
+    g_pCore->GetGUI()->ResetCursorColor(255.f, 255.f, 255.f, 255.f);
```
If the function is meant to reset the color to the default, it should not take any arguments and should simply set the color to the default, which is white, by default

Proxy-99 committed:
remove setAlpha

Proxy-99 committed:
part 2

fix noexcept and Hungarian notation

Proxy-99 committed:
override and fixes

Proxy-99 committed:
Merge branch 'master' into cursor

FileEX commented on Client/gui/CGUI_Impl.h:
```diff
@@ -311,7 +316,10 @@ class CGUI_Impl : public CGUI, public CGUITabList
+    float                 CurrentServerCursorAlpha = 255.0f;
+    float                 CurrentServerCursorRed = 255.0f;
+    float                 CurrentServerCursorGreen = 255.0f;
+    float                 CurrentServerCursorBlue = 255.0f;
```
```cpp
CurrentServerCursorAlpha{255.0f};
// etc
```

FileEX commented on Client/mods/deathmatch/logic/luadefs/CLuaClientDefs.cpp:
```diff
@@ -80,3 +82,20 @@ bool CLuaClientDefs::IsCapsLockEnabled()
+
+bool CLuaClientDefs::SetCursorColor(std::optional<float> r, std::optional<float> g, std::optional<float> b, std::optional<float> alpha) noexcept
+{
+    if (!g_pCore->IsMenuVisible()) 
```
IsMenuVisible isn't marked as noexcept

Proxy-99 committed:
review

Proxy-99 committed:
Merge branch 'cursor' of https://github.com/Proxy-99/mtasa-blue into cursor

Proxy-99 committed:
Merge branch 'master' into cursor

Proxy-99 commented:
> In some skins, even the default ones, the cursor isn't white, so using a white cursor would either look out of place or be completely irrelevant

I test it with others themes it works normally

Fernando-A-Rocha commented:
> > In some skins, even the default ones, the cursor isn't white, so using a white cursor would either look out of place or be completely irrelevant
> 
> I test it with others themes it works normally

Would it be possible to reset the color to the color set by the current theme?

Proxy-99 committed:
Merge branch 'master' into cursor

FileEX commented:
Please, resolve conflicts

Proxy-99 committed:
Merge branch 'master' into cursor

Proxy-99 commented:
> Please, resolve conflicts

@FileEX done

tederis commented:
@Proxy-99 Please, resolve conflicts

Proxy-99 committed:
Merge branch 'master' into cursor

Proxy-99 committed:
empty line
